cmd/maya-apiserver/cstor-operator/cspc/handler.go

```diff
@@ -0,0 +1,162 @@
+// with the current status of the resource.
+func (c *Controller) syncHandler(key string) error {
+	startTime := time.Now()
+	glog.V(4).Infof("Started syncing storagepoolclaim %q (%v)", key, startTime)
+	defer func() {
+		glog.V(4).Infof("Finished syncing storagepoolclaim %q (%v)", key, time.Since(startTime))
+	}()
+
+	// Convert the namespace/name string into a distinct namespace and name
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	// Get the cspc resource with this namespace/name
+	cspc, err := c.cspcLister.CStorPoolClusters(ns).Get(name)
+	if k8serror.IsNotFound(err) {
+		runtime.HandleError(fmt.Errorf("cspc '%s' has been deleted", key))
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	// Deep-copy otherwise we are mutating our cache.
+	// TODO: Deep-copy only when needed.
+	cspcGot := cspc.DeepCopy()
+	err = c.syncCSPC(cspcGot)
+	return err
+}
+
+// enqueueSpc takes a SPC resource and converts it into a namespace/name
+// string which is then put onto the work queue. This method should *not* be
+// passed resources of any type other than SPC.
+func (c *Controller) enqueueSpc(cspc interface{}) {
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(cspc); err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	c.workqueue.Add(key)
+}
+
+// synSpc is the function which tries to converge to a desired state for the cspc.
+func (c *Controller) syncCSPC(cspc *apis.CStorPoolCluster) error {
+	//err := validate(cspc)
+	//if err != nil {
+	//	glog.Errorf("Validation of cspc failed:%s", err)
+	//	return nil
+	//}
+	openebsNameSpace := env.Get(env.OpenEBSNamespace)
+	if openebsNameSpace == "" {
+		return errors.Errorf("Could not sync CSPC {%s}: got empty namespace for openebs from env variable", cspc.Name)
+	}
+	pc, err := c.NewPoolConfig(cspc, openebsNameSpace)
+	if err != nil {
+		return errors.Wrapf(err, "Could not sync CSPC {%s}: failed to get pool config", cspc.Name)
+	}
+	pendingPoolCount, err := pc.AlgorithConfig.GetPendingPoolCount()
+	if err != nil {
+		return err
+	}
+	if pendingPoolCount > 0 {
+		err = pc.create(pendingPoolCount, cspc)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// create is a wrapper function that calls the actual function to create pool as many time
+// as the number of pools need to be created.
+func (pc *PoolConfig) create(pendingPoolCount int, cspc *apis.CStorPoolCluster) error {
+	var newSpcLease Leaser
+	newSpcLease = &Lease{cspc, CSPCLeaseKey, pc.Controller.clientset, pc.Controller.kubeclientset}
+	err := newSpcLease.Hold()
+	if err != nil {
+		return errors.Wrapf(err, "Could not acquire lease on cspc object")
+	}
+	glog.V(4).Infof("Lease acquired successfully on storagepoolclaim %s ", cspc.Name)
+	defer newSpcLease.Release()
+	for poolCount := 1; poolCount <= pendingPoolCount; poolCount++ {
```

Comment:
better to get the pendingPoolCount again.. reason is:
- if there are two threads which are executing syncCSPC and both found that pendingPoolCount is > 0
- both of them enters this function `create`. One will get the hold and created pendingPoolCount number of replicas. Other thread also will get the Hold and tries to create pendingPoolCount number of replicas again

Reply:
This issue won't happen for multiple apiservers and we will take it later for multi thread thing.